xarray/core/variable.py

```diff
@@ -404,34 +416,189 @@ def _item_key_to_tuple(self, key):
+                dims.append(k.dims[0])
+            if not isinstance(k, integer_types):
+                dims.append(d)
+
+        if len(set(dims)) == len(dims):
+            return self._broadcast_indexes_outer(key)
+
+        return self._broadcast_indexes_vectorized(key)
+
+    def _broadcast_indexes_basic(self, key):
+        dims = tuple(dim for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        return dims, BasicIndexer(key), None
+
+    def _validate_indexers(self, key):
+        """ Make sanity checks """
+        for dim, k in zip(self.dims, key):
+            if isinstance(k, BASIC_INDEXING_TYPES):
+                pass
+            else:
+                if not isinstance(k, Variable):
+                    k = np.asarray(k)
+                    if k.ndim > 1:
+                        raise IndexError(
+                            "Unlabeled multi-dimensional array cannot be "
+                            "used for indexing: {}".format(k))
+                if k.dtype.kind == 'b':
+                    if self.shape[self.get_axis_num(dim)] != len(k):
+                        raise IndexError(
+                            "Boolean array size {0:d} is used to index array "
+                            "with shape {1:s}.".format(len(k),
+                                                       str(self.shape)))
+                    if k.ndim > 1:
+                        raise IndexError("{}-dimensional boolean indexing is "
+                                         "not supported. ".format(k.ndim))
+                    if getattr(k, 'dims', (dim, )) != (dim, ):
+                        raise IndexError(
+                            "Boolean indexer should be unlabeled or on the "
+                            "same dimension to the indexed array. Indexer is "
+                            "on {0:s} but the target dimension is "
+                            "{1:s}.".format(str(k.dims), dim))
+
+    def _broadcast_indexes_outer(self, key):
+        dims = tuple(k.dims[0] if isinstance(k, Variable) else dim
+                     for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        indexer = []
+        for k in key:
+            if isinstance(k, Variable):
+                k = k.data
+
+            if isinstance(k, BASIC_INDEXING_TYPES):
+                indexer.append(k)
+            else:
+                k = np.asarray(k)
+                indexer.append(k if k.dtype.kind != 'b' else np.flatnonzero(k))
+        return dims, OuterIndexer(indexer), None
+
+    def _nonzero(self):
+        """ Equivalent numpy's nonzero but returns a tuple of Varibles. """
+        # TODO we should replace dask's native nonzero
+        # after https://github.com/dask/dask/issues/1076 is implemented.
+        nonzeros = np.nonzero(self.data)
+        return tuple(Variable((dim), nz) for nz, dim
+                     in zip(nonzeros, self.dims))
```

Comment:
So this is in dask now. Can we add a conditional import of the dask version here?

Side note, I think a nonzero method just like this would fit in xarray's public API (not part of this PR).